File: C++/9-DynamicProgramming(DP)/Problems/4-LongestPalindromicSubstring/LongestPalindromicSubstring.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <functional> // Include this for std::function

using namespace std;

// Function to print the Longest Palindromic Substring and its length
void printResult(const string& s, int start, int maxLength) {
    cout << "Longest Palindromic Substring: ";
    for (int i = start; i < start + maxLength; ++i) {
        cout << s[i];
    }
    cout << endl;
    cout << "Length: " << maxLength << endl;
}

// 1. Dynamic Programming - Tabulation (Bottom-Up)
// Time Complexity: O(n^2), Space Complexity: O(n^2)
string longestPalindromeDP(const string& s) {
    int n = s.size();
    if (n <= 1) return s;

    // Create a table to store whether a substring s[i...j] is a palindrome
    vector<vector<bool>> dp(n, vector<bool>(n, false));

    // All single characters are palindromes
    for (int i = 0; i < n; ++i) {
        dp[i][i] = true;
    }

    // Check for palindromes of length 2
    int maxLength = 1;
    int start = 0;
    for (int i = 0; i < n - 1; ++i) {
        if (s[i] == s[i + 1]) {
            dp[i][i + 1] = true;
            start = i;
            maxLength = 2;
        }
    }

    // Check for palindromes of length greater than 2
    for (int k = 3; k <= n; ++k) {
        for (int i = 0; i <= n - k; ++i) {
            int j = i + k - 1; // Ending index of the substring
            if (s[i] == s[j] && dp[i + 1][j - 1]) {
                dp[i][j] = true;
                if (k > maxLength) {
                    start = i;
                    maxLength = k;
                }
            }
        }
    }
    return s.substr(start, maxLength);
}

// 2. Optimized Dynamic Programming - Tabulation (Bottom-Up)
// Time Complexity: O(n^2), Space Complexity: O(n^2) - but often faster in practice due to simpler logic
string longestPalindromeDPOptimized(const string& s) {
    int n = s.size();
    if (n <= 1) return s;

    vector<vector<bool>> dp(n, vector<bool>(n, false));
    int maxLength = 1;
    int start = 0;

    for (int i = n - 1; i >= 0; --i) { // Iterate from the end
        for (int j = i; j < n; ++j) {
            if (s[i] == s[j] && (j - i <= 2 || dp[i + 1][j - 1])) {
                dp[i][j] = true;
                if (j - i + 1 > maxLength) {
                    maxLength = j - i + 1;
                    start = i;
                }
            }
        }
    }
    return s.substr(start, maxLength);
}

// 3. Expand Around Center Approach
// Time Complexity: O(n^2), Space Complexity: O(1)
string longestPalindromeExpand(const string& s) {
    int n = s.size();
    if (n <= 1) return s;

    int start = 0;
    int maxLength = 1;

    for (int i = 0; i < n; ++i) {
        // For odd length palindromes
        int left = i, right = i;
        while (left >= 0 && right < n && s[left] == s[right]) {
            if (right - left + 1 > maxLength) {
                start = left;
                maxLength = right - left + 1;
            }
            --left;
            ++right;
        }

        // For even length palindromes
        left = i, right = i + 1;
        while (left >= 0 && right < n && s[left] == s[right]) {
            if (right - left + 1 > maxLength) {
                start = left;
                maxLength = right - left + 1;
            }
            --left;
            ++right;
        }
    }
    return s.substr(start, maxLength);
}

// 4. Manacher's Algorithm
// Time Complexity: O(n), Space Complexity: O(n)
// This is the most efficient algorithm, linear time complexity
string longestPalindromeManacher(const string& s) {
    int n = s.size();
    if (n <= 1) return s;

    // Preprocess the string to handle even length palindromes
    string processedString = "^";  // Start and end markers to avoid boundary checks
    for (char c : s) {
        processedString += c;
        processedString += "#";
    }
    processedString += "$";

    int processedLength = processedString.size();
    vector<int> p(processedLength, 0); // Array to store palindrome lengths at each center
    int center = 0, right = 0;
    int maxLen = 0, start = 0;

    for (int i = 1; i < processedLength - 1; ++i) {
        int mirror = 2 * center - i; // Mirror position of i with respect to center

        if (i < right) {
            p[i] = min(right - i, p[mirror]); // Take minimum of mirror's length and right boundary
        }

        // Expand around center i
        while (processedString[i + (1 + p[i])] == processedString[i - (1 + p[i])]) {
            p[i]++;
        }

        // If current palindrome expands past right, update center and right
        if (i + p[i] > right) {
            center = i;
            right = i + p[i];
        }

        // Update maxLen and start
        if (p[i] > maxLen) {
            maxLen = p[i];
            start = (i - maxLen) / 2; // Original start position
        }
    }
    return s.substr(start, maxLen);
}

// 5. Recursive Approach with Memoization (Top-Down Dynamic Programming)
// Time Complexity: O(n^2), Space Complexity: O(n^2)
string longestPalindromeRecursive(const string& s) {
    int n = s.size();
    if (n <= 1) return s;

    // Memoization table
    vector<vector<int>> memo(n, vector<int>(n, -1));
    int start = 0;
    int maxLength = 1;

    std::function<bool(int, int)> isPalindrome = [&](int i, int j) {
        if (i > j) return true;
        if (memo[i][j] != -1) return memo[i][j] == 1; // Return memoized result as bool
        // Removed the else condition and simplified
        if (s[i] == s[j]) {
            memo[i][j] = isPalindrome(i + 1, j - 1) ? 1 : 0;
        }
        else{
             memo[i][j] = 0;
        }
        return memo[i][j] == 1;
    };

    for (int i = 0; i < n; ++i) {
        for (int j = i; j < n; ++j) {
            if (isPalindrome(i, j) && (j - i + 1 > maxLength)) {
                maxLength = j - i + 1;
                start = i;
            }
        }
    }
    return s.substr(start, maxLength);
}

int main() {
    string s = "babad"; // Example string
    cout << "Input String: " << s << endl << endl;

    // 1. Dynamic Programming - Tabulation
    cout << "1. Dynamic Programming - Tabulation:" << endl;
    string resultDP = longestPalindromeDP(s);
    cout << "Longest Palindromic Substring: " << resultDP << endl;

    // 2. Optimized Dynamic Programming - Tabulation
    cout << "\n2. Optimized Dynamic Programming - Tabulation:" << endl;
    string resultDPOptimized = longestPalindromeDPOptimized(s);
    cout << "Longest Palindromic Substring: " << resultDPOptimized << endl;

    // 3. Expand Around Center
    cout << "\n3. Expand Around Center:" << endl;
    string resultExpand = longestPalindromeExpand(s);
    cout << "Longest Palindromic Substring: " << resultExpand << endl;

    // 4. Manacher's Algorithm
    cout << "\n4. Manacher's Algorithm:" << endl;
    string resultManacher = longestPalindromeManacher(s);
    cout << "Longest Palindromic Substring: " << resultManacher << endl;

    // 5. Recursive with Memoization
    cout << "\n5. Recursive Approach with Memoization:" << endl;
    string resultRecursive = longestPalindromeRecursive(s);
    cout << "Longest Palindromic Substring: " << resultRecursive << endl;

    return 0;
}

File: Java/9-DynamicProgramming(DP)/Problems/4-LongestPalindromicSubstring/LongestPalindromicSubstring.java
import java.util.Arrays;

public class LongestPalindromicSubstring {

    /**
     * Approach 1: Brute Force
     * -   Time Complexity: O(n^3)
     * -   Space Complexity: O(1)
     *
     * Iterates through all possible substrings and checks if each one is a palindrome.
     */
    public static String longestPalindromeBruteForce(String s) {
        if (s == null || s.length() < 1) return "";
        String longest = "";
        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                String sub = s.substring(i, j + 1);
                if (isPalindrome(sub) && sub.length() > longest.length()) {
                    longest = sub;
                }
            }
        }
        return longest;
    }

    /**
     * Helper function to check if a string is a palindrome.
     */
    private static boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    /**
     * Approach 2: Dynamic Programming
     * -   Time Complexity: O(n^2)
     * -   Space Complexity: O(n^2)
     *
     * Creates a 2D table where dp[i][j] is true if the substring s[i...j] is a palindrome.
     */
    public static String longestPalindromeDP(String s) {
        if (s == null || s.length() < 1) return "";
        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int start = 0;
        int maxLen = 1;

        // Base case: single characters are palindromes
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        // Base case: two-character substrings
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLen = 2;
            }
        }

        // Check for palindromes of length 3 and greater
        for (int len = 3; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    if (len > maxLen) {
                        start = i;
                        maxLen = len;
                    }
                }
            }
        }
        return s.substring(start, start + maxLen);
    }

    /**
     * Approach 3: Optimized Dynamic Programming
     * -   Time Complexity: O(n^2)
     * -   Space Complexity: O(n^2)
     *
     * Same as DP, but with a slightly different way of filling the table.
     */
    public static String longestPalindromeDPOptimized(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int maxLen = 1;
        int start = 0;

        for (int j = 0; j < n; j++) {
            for (int i = 0; i <= j; i++) {
                if (s.charAt(i) == s.charAt(j) && (j - i <= 2 || dp[i + 1][j - 1])) {
                    dp[i][j] = true;
                    if (j - i + 1 > maxLen) {
                        maxLen = j - i + 1;
                        start = i;
                    }
                }
            }
        }
        return s.substring(start, start + maxLen);
    }



    /**
     * Approach 4: Expand Around Center
     * -   Time Complexity: O(n^2)
     * -   Space Complexity: O(1)
     *
     * Considers each character as a potential center of a palindrome and expands around it.
     */
    public static String longestPalindromeExpandAroundCenter(String s) {
        if (s == null || s.length() < 1) return "";
        int start = 0;
        int maxLen = 1;
        for (int i = 0; i < s.length(); i++) {
            // For odd length palindromes
            int len1 = expandAroundCenter(s, i, i);
            // For even length palindromes
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > maxLen) {
                maxLen = len;
                start = i - (len - 1) / 2;
            }
        }
        return s.substring(start, start + maxLen);
    }

    /**
     * Helper function to expand around the center of a potential palindrome.
     */
    private static int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    /**
     * Approach 5: Manacher's Algorithm
     * -   Time Complexity: O(n)
     * -   Space Complexity: O(n)
     *
     * The most efficient approach.  Transforms the string to handle even/odd lengths,
     * and uses an array to store palindrome lengths.
     */
    public static String longestPalindromeManacher(String s) {
        if (s == null || s.length() < 1) return "";

        // Preprocess the string to handle even and odd length palindromes
        String processedString = preprocess(s); // O(n)
        int n = processedString.length();
        int[] p = new int[n]; // Array to store palindrome lengths
        int center = 0, right = 0;
        int maxLen = 0, maxCenter = 0;

        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * center - i;  // Get the mirror of i around the center.

            if (i < right) {
                p[i] = Math.min(right - i, p[mirror]); // Take min of right bound and mirror's p.
            }

            // Expand around center i.
            while (processedString.charAt(i + (1 + p[i])) == processedString.charAt(i - (1 + p[i]))) {
                p[i]++;
            }

            // If current palindrome expands past right, adjust center and right.
            if (i + p[i] > right) {
                center = i;
                right = i + p[i];
            }

            // Track the maximum palindrome length.
            if (p[i] > maxLen) {
                maxLen = p[i];
                maxCenter = i;
            }
        }
        // Extract the original longest palindrome substring.
        int start = (maxCenter - maxLen) / 2;
        return s.substring(start, start + maxLen);
    }

    /**
     * Helper function to preprocess the string for Manacher's Algorithm.
     * Inserts '#' between characters and at the beginning and end.
     * Example: "aba" -> "^#a#b#a#$"
     */
    private static String preprocess(String s) {
        if (s == null || s.length() == 0) return "^$"; // Handle empty string
        StringBuilder sb = new StringBuilder("^");
        for (int i = 0; i < s.length(); i++) {
            sb.append("#").append(s.charAt(i));
        }
        sb.append("#$");
        return sb.toString();
    }

    public static void main(String[] args) {
        String s = "babad";
        //String s = "cbbd";
        //String s = "a";
        //String s = "ac";

        System.out.println("Input: " + s);
        System.out.println("Longest Palindromic Substring (Brute Force): " + longestPalindromeBruteForce(s));
        System.out.println("Longest Palindromic Substring (Dynamic Programming): " + longestPalindromeDP(s));
        System.out.println("Longest Palindromic Substring (DP Optimized): " + longestPalindromeDPOptimized(s));
        System.out.println("Longest Palindromic Substring (Expand Around Center): " + longestPalindromeExpandAroundCenter(s));
        System.out.println("Longest Palindromic Substring (Manacher's Algorithm): " + longestPalindromeManacher(s));
    }
}

